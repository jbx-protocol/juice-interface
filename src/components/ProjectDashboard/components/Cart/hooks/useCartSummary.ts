import { useProjectCart } from 'components/ProjectDashboard/hooks'
import { NftRewardsContext } from 'contexts/NftRewards/NftRewardsContext'
import { useWallet } from 'hooks/Wallet'
import { useCallback, useContext, useMemo } from 'react'
import { formatCurrencyAmount } from 'utils/formatCurrencyAmount'

export const useCartSummary = () => {
  const { isConnected, connect } = useWallet()
  const cart = useProjectCart()
  const nftRewardsCtx = useContext(NftRewardsContext).nftRewards
  const tiers = useMemo(
    () => nftRewardsCtx.rewardTiers ?? [],
    [nftRewardsCtx.rewardTiers],
  )
  const nftRewards = useMemo(() => {
    return cart.nftRewards
      .map(nft => ({ ...tiers.find(tier => tier.id === nft.id), ...nft }))
      .filter(nft => nft !== undefined)
      .map(nft => ({
        fileUrl: nft.fileUrl,
        name: nft.name,
        quantity: nft.quantity,
        id: nft.id,
      }))
  }, [cart.nftRewards, tiers])

  const amountText = cart.totalAmount
    ? formatCurrencyAmount(cart.totalAmount)
    : undefined

  const resetCart = useCallback(() => {
    cart.dispatch({ type: 'reset' })
  }, [cart])

  const payProject = useCallback(() => {
    if (!isConnected) {
      connect()
      return
    }
    cart.dispatch({ type: 'openPayModal' })
  }, [cart, connect, isConnected])

  const showCurrencyOnCollapse = useMemo(() => {
    return cart.payAmount?.amount ? cart.payAmount.amount > 0 : false
  }, [cart.payAmount])

  return {
    amountText,
    currency: cart.totalAmount?.currency,
    nftRewards,
    walletConnected: isConnected,
    showCurrencyOnCollapse,
    resetCart,
    payProject,
  }
}
