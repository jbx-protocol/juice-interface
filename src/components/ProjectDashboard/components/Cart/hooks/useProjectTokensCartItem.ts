import {
  useProjectCart,
  useTokensPerEth,
} from 'components/ProjectDashboard/hooks'
import { useCallback } from 'react'

export const useProjectTokensCartItem = () => {
  const {
    totalAmount: payAmount,
    userIsReceivingTokens,
    dispatch,
  } = useProjectCart()
  const { receivedTickets } = useTokensPerEth(payAmount)

  const removeTokens = useCallback(
    () => dispatch({ type: 'removeTokens' }),
    [dispatch],
  )
  return {
    userIsReceivingTokens,
    receivedTickets,
    removeTokens,
  }
}
