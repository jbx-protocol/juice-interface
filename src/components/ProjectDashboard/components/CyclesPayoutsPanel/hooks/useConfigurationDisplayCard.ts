import { t } from '@lingui/macro'
import { useMemo } from 'react'

export const useConfigurationDisplayCard = (type: 'current' | 'upcoming') => {
  const title = useMemo(() => {
    if (type === 'current') {
      return t`Current cycle`
    }
    return t`Upcoming cycle`
  }, [type])

  return {
    title,
  }
}
