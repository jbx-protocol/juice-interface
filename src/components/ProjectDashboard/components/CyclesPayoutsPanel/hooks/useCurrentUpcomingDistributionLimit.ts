import {
  useProjectContext,
  useProjectMetadata,
} from 'components/ProjectDashboard/hooks'
import useProjectDistributionLimit from 'hooks/v2v3/contractReader/useProjectDistributionLimit'
import { useProjectUpcomingFundingCycle } from 'hooks/v2v3/contractReader/useProjectUpcomingFundingCycle'

export const useCurrentUpcomingDistributionLimit = (
  type: 'current' | 'upcoming',
) => {
  const { projectId } = useProjectMetadata()
  const {
    primaryETHTerminal,
    distributionLimit: currentDistributionLimit,
    distributionLimitCurrency: currentDistributionLimitCurrency,
  } = useProjectContext()

  const {
    data: upcomingFundingCycleData,
    loading: upcomingFundingCycleLoading,
  } = useProjectUpcomingFundingCycle({ projectId })
  const [upcomingFundingCycle] = upcomingFundingCycleData ?? []

  const {
    data: upcomingDistributionLimitData,
    loading: upcomingDistributionLimitLoading,
  } = useProjectDistributionLimit({
    projectId,
    configuration: upcomingFundingCycle?.configuration.toString(),
    terminal: primaryETHTerminal,
  })

  const [upcomingDistributionLimit, upcomingDistributionLimitCurrency] =
    upcomingDistributionLimitData ?? []

  if (type === 'current') {
    return {
      distributionLimit: currentDistributionLimit,
      distributionLimitCurrency: currentDistributionLimitCurrency,
      loading: false,
    }
  }
  return {
    distributionLimit: upcomingDistributionLimit,
    distributionLimitCurrency: upcomingDistributionLimitCurrency,
    loading: upcomingFundingCycleLoading || upcomingDistributionLimitLoading,
  }
}
