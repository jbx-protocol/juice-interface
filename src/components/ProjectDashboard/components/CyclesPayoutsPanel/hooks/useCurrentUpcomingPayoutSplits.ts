import {
  useProjectContext,
  useProjectMetadata,
} from 'components/ProjectDashboard/hooks'
import { ETH_PAYOUT_SPLIT_GROUP } from 'constants/splits'
import useProjectSplits from 'hooks/v2v3/contractReader/useProjectSplits'
import { useProjectUpcomingFundingCycle } from 'hooks/v2v3/contractReader/useProjectUpcomingFundingCycle'

export const useCurrentUpcomingPayoutSplits = (
  type: 'current' | 'upcoming',
) => {
  const { projectId } = useProjectMetadata()
  const { payoutSplits: currentPayoutSplits } = useProjectContext()

  const {
    data: upcomingFundingCycleData,
    loading: upcomingFundingCycleLoading,
  } = useProjectUpcomingFundingCycle({ projectId })
  const [upcomingFundingCycle] = upcomingFundingCycleData ?? []

  const { data: upcomingPayoutSplits, loading: upcomingProjectSplitsLoading } =
    useProjectSplits({
      projectId,
      splitGroup: ETH_PAYOUT_SPLIT_GROUP,
      domain: upcomingFundingCycle?.configuration?.toString(),
    })

  if (type === 'current') {
    return { splits: currentPayoutSplits, loading: false }
  }
  return {
    splits: upcomingPayoutSplits,
    loading: upcomingFundingCycleLoading || upcomingProjectSplitsLoading,
  }
}
