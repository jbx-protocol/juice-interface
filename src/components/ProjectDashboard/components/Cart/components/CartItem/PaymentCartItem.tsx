import { CurrencyIcon } from 'components/ProjectDashboard/components/ui/CurrencyIcon'
import { useProjectCart } from 'components/ProjectDashboard/hooks'
import { handleConfirmationDeletion } from 'components/ProjectDashboard/utils/modals'
import { useCallback } from 'react'
import { CartItem } from './CartItem'

export const PaymentCartItem = () => {
  const { payAmount, dispatch } = useProjectCart()

  const removePayment = useCallback(() => {
    dispatch({ type: 'removePayment' })
  }, [dispatch])

  if (!payAmount || payAmount.amount === 0) {
    return null
  }

  return (
    <CartItem
      title={
        <span className="flex items-center gap-2">
          <span>Payment</span>
        </span>
      }
      icon={
        <CurrencyIcon
          innerIconClassName="h-8 w-8"
          className="h-14 w-14"
          currency={payAmount.currency}
        />
      }
      price={payAmount}
      onRemove={handleConfirmationDeletion({
        type: 'payment',
        onConfirm: removePayment,
      })}
    />
  )
}
