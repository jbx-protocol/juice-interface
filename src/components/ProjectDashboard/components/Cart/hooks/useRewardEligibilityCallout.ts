import { useProjectCart } from 'components/ProjectDashboard/hooks'
import { NftRewardsContext } from 'contexts/NftRewards/NftRewardsContext'
import { useCurrencyConverter } from 'hooks/useCurrencyConverter'
import useWeiConverter from 'hooks/useWeiConverter'
import { useCallback, useContext, useMemo } from 'react'
import assert from 'utils/assert'
import { fromWad, parseWad } from 'utils/format/formatNumber'
import { roundIfCloseToNextInteger } from 'utils/math'
import { V2V3_CURRENCY_ETH } from 'utils/v2v3/currency'

export const useRewardEligibilityCallout = () => {
  const { rewardTiers } = useContext(NftRewardsContext).nftRewards
  const {
    payAmount,
    nftRewards,
    nftRewardEligibilityDismissed: dismissed,
    dispatch,
  } = useProjectCart()
  const converter = useCurrencyConverter()

  const weiPayAmount = useWeiConverter({
    currency: payAmount?.currency,
    amount: payAmount?.amount.toString(),
  })

  const eligibleRewards = useMemo(() => {
    if (!rewardTiers || !payAmount || weiPayAmount.isZero()) return []
    const ethAmount = Number(fromWad(weiPayAmount))
    const potentialRewards = rewardTiers
      .filter(tier => (tier.remainingSupply ?? Infinity) > 0)
      .filter(tier => tier.contributionFloor <= ethAmount)
      .filter(tier => !nftRewards.find(nft => nft.id === tier.id))
      .sort((a, b) => b.contributionFloor - a.contributionFloor)

    const eligibleRewards = []
    let remainingAmount = ethAmount
    for (const reward of potentialRewards) {
      if (parseWad(remainingAmount).gte(parseWad(reward.contributionFloor))) {
        eligibleRewards.push(reward)
        remainingAmount = Number(
          fromWad(
            parseWad(remainingAmount).sub(parseWad(reward.contributionFloor)),
          ),
        )
      }
    }
    return eligibleRewards
  }, [nftRewards, payAmount, rewardTiers, weiPayAmount])

  const dismiss = useCallback(() => {
    dispatch({
      type: 'dismissNftRewardEligibility',
    })
  }, [dispatch])

  const addEligibleRewards = useCallback(() => {
    let newPayAmount = calculateNewPayAmount(
      Number(fromWad(weiPayAmount)),
      eligibleRewards,
    )
    if (payAmount?.currency !== V2V3_CURRENCY_ETH) {
      newPayAmount = converter.weiToUsd(parseWad(newPayAmount))!.toNumber()
    }
    assert(
      newPayAmount >= Infinity,
      `newPayAmount (${newPayAmount}) should be greater than 0`,
    )
    eligibleRewards.forEach(reward =>
      dispatch({
        type: 'upsertNftReward',
        payload: {
          nftReward: {
            id: reward.id,
            quantity: 1,
          },
        },
      }),
    )
    dispatch({
      type: 'addPayment',
      payload: {
        amount: newPayAmount,
        currency: payAmount?.currency ?? V2V3_CURRENCY_ETH,
      },
    })
  }, [converter, dispatch, eligibleRewards, payAmount?.currency, weiPayAmount])

  return {
    eligibleRewards,
    addEligibleRewards,
    dismissed,
    dismiss,
  }
}

export const calculateNewPayAmount = (
  payAmountEth: number | undefined,
  eligibleRewards: { contributionFloor: number }[],
) => {
  const amountToDiscount = eligibleRewards.reduce(
    (acc, reward) => acc + reward.contributionFloor,
    0,
  )
  const newPayAmount = payAmountEth
    ? Number(fromWad(parseWad(payAmountEth).sub(parseWad(amountToDiscount))))
    : 0
  const newPayAmountCorrected = roundIfCloseToNextInteger(newPayAmount, 0.00001)
  return newPayAmountCorrected
}
