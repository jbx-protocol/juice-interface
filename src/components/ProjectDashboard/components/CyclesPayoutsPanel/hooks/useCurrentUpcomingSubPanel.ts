import { t } from '@lingui/macro'
import {
  useProjectContext,
  useProjectMetadata,
} from 'components/ProjectDashboard/hooks'
import { useFundingCycleCountdown } from 'components/ProjectDashboard/hooks/useFundingCycleCountdown'
import { timeSecondsToDateString } from 'components/ProjectDashboard/utils/timeSecondsToDateString'
import { useProjectUpcomingFundingCycle } from 'hooks/v2v3/contractReader/useProjectUpcomingFundingCycle'
import { useMemo } from 'react'

export const useCurrentUpcomingSubPanel = (type: 'current' | 'upcoming') => {
  const { projectId } = useProjectMetadata()
  const {
    fundingCycle,
    loading: { fundingCycleLoading },
  } = useProjectContext()
  const {
    data: upcomingFundingCycleData,
    loading: upcomingFundingCycleLoading,
  } = useProjectUpcomingFundingCycle({ projectId })
  const [upcomingFundingCycle] = upcomingFundingCycleData ?? []
  const { timeRemainingText } = useFundingCycleCountdown()

  const cycleNumber = useMemo(() => {
    if (type === 'current') {
      return fundingCycle?.number.toNumber()
    }
    return fundingCycle?.number ? fundingCycle.number.toNumber() + 1 : undefined
  }, [fundingCycle?.number, type])

  const cycleUnlocked = useMemo(() => {
    if (type === 'current') {
      return fundingCycle?.duration?.isZero() ?? true
    }
    return upcomingFundingCycle?.duration?.isZero() ?? true
  }, [fundingCycle?.duration, type, upcomingFundingCycle?.duration])

  const upcomingCycleLength = useMemo(() => {
    if (!upcomingFundingCycle) return
    if (cycleUnlocked) return '-'
    return timeSecondsToDateString(
      upcomingFundingCycle.duration.toNumber(),
      'short',
    )
  }, [cycleUnlocked, upcomingFundingCycle])

  const status = useMemo(() => {
    if (cycleUnlocked) return t`Unlocked`
    return t`Locked`
  }, [cycleUnlocked])

  const remainingTime = useMemo(() => {
    if (cycleUnlocked) return '-'
    return timeRemainingText
  }, [cycleUnlocked, timeRemainingText])

  // Short circuit current for faster loading
  if (type === 'current') {
    if (fundingCycleLoading) return { loading: true, type }
    return {
      loading: false,
      type,
      cycleNumber,
      status,
      remainingTime,
    }
  }

  if (fundingCycleLoading || upcomingFundingCycleLoading)
    return {
      loading: true,
      type,
    }

  return {
    loading: false,
    type,
    cycleNumber,
    status,
    cycleLength: upcomingCycleLength,
  }
}
