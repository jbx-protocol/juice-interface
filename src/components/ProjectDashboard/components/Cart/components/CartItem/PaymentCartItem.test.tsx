/**
 * @jest-environment jsdom
 */
import { render } from '@testing-library/react'
import { useProjectCart } from 'components/ProjectDashboard/hooks'
import { V2V3_CURRENCY_ETH } from 'utils/v2v3/currency'
import { PaymentCartItem } from './PaymentCartItem'

jest.mock('components/ProjectDashboard/hooks')

describe('PaymentCartItem', () => {
  const DefaultUseProjectCartMock = {
    dispatch: jest.fn(),
    payAmount: {
      amount: 100,
      currency: V2V3_CURRENCY_ETH,
    },
    expanded: false,
    visible: true,
  }
  beforeEach(() => {
    ;(useProjectCart as jest.Mock).mockReturnValue(DefaultUseProjectCartMock)
  })

  it('renders', () => {
    const { container } = render(<PaymentCartItem />)
    expect(container).toMatchSnapshot()
  })
})
