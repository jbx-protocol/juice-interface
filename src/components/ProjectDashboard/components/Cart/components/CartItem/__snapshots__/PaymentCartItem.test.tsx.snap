// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaymentCartItem renders 1`] = `
<div>
  <div
    class="flex min-w-0 items-center justify-between gap-4 border-b border-grey-200 py-5 dark:border-slate-500"
  >
    <div
      class="flex min-w-0 items-center"
    >
      <div
        class="flex items-center justify-center rounded-full p-1 transition-colors bg-bluebs-50 text-bluebs-500 dark:bg-bluebs-500 dark:text-bluebs-950 h-14 w-14"
        data-testid="pay-input-currency-icon"
      >
        <svg
          class="h-8 w-8"
          fill="none"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M18.9828 11.8686C19 11.878 19 11.8968 19 11.9155L19.0086 11.8968V11.9249V11.9437C19 11.9531 19 11.9718 19 11.9718C18.9914 11.9812 18.9914 11.9906 18.9914 11.9906C18.9828 12 18.9656 12.0188 18.9571 12.0188H18.9399L12.0687 15.773C12.0515 15.7918 12.0344 15.7918 12.0086 15.7918C11.9828 15.7918 11.9485 15.773 11.9485 15.773L5.0773 12.0188H5.06012C5.04294 12.0094 5.03436 12 5.02577 11.9906H5.01718V11.9812H5.00859C5 11.9718 5 11.9531 5 11.9531V11.9343V11.9061H5.00859C5.00859 11.8924 5.01319 11.8837 5.01903 11.8726C5.02116 11.8686 5.02347 11.8642 5.02577 11.8592V11.8498L11.8883 1.04693L11.9055 1.02816L11.9055 1.02815C11.9141 1.01877 11.9313 1 11.9399 1H11.9571H12C12.0086 1 12.0429 1.00939 12.0429 1.00939H12.0515C12.0687 1.01877 12.0773 1.02816 12.0859 1.03754L12.0859 1.03754L12.1031 1.05631L18.9828 11.8592V11.8686ZM12.0945 17.9505L18.3301 14.3464H18.3129C18.3816 14.3089 18.4589 14.3276 18.5018 14.3933C18.5534 14.459 18.5448 14.5529 18.5018 14.6092L12.1031 22.9437C12.0773 22.9812 12.0344 23 11.9914 23C11.9485 23 11.8798 22.9437 11.8798 22.9437L5.48098 14.6092C5.43804 14.5435 5.43804 14.459 5.48098 14.3933C5.53252 14.3276 5.60982 14.3089 5.66994 14.3464L11.9055 17.9505C11.9656 17.9881 12.0344 17.9881 12.0945 17.9505Z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
      </div>
      <span
        class="ml-3 min-w-0 font-medium dark:text-slate-50"
      >
        <span
          class="flex items-center gap-2"
        >
          <span>
            Payment
          </span>
        </span>
      </span>
    </div>
    <div
      class="flex items-center"
    >
      <span
        class="mr-4 text-sm font-medium"
      >
        Îž100
      </span>
      <svg
        aria-hidden="true"
        class="inline h-6 w-6 text-grey-400 dark:text-slate-300 md:h-4 md:w-4"
        data-testid="cart-item-remove-button"
        fill="none"
        role="button"
        stroke="currentColor"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M14.74 9l-.346 9m-4.788 0L9.26 9m9.968-3.21c.342.052.682.107 1.022.166m-1.022-.165L18.16 19.673a2.25 2.25 0 01-2.244 2.077H8.084a2.25 2.25 0 01-2.244-2.077L4.772 5.79m14.456 0a48.108 48.108 0 00-3.478-.397m-12 .562c.34-.059.68-.114 1.022-.165m0 0a48.11 48.11 0 013.478-.397m7.5 0v-.916c0-1.18-.91-2.164-2.09-2.201a51.964 51.964 0 00-3.32 0c-1.18.037-2.09 1.022-2.09 2.201v.916m7.5 0a48.667 48.667 0 00-7.5 0"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </div>
  </div>
</div>
`;
